Zamg weather

Zac-HD opened the pull request:
**Description:** improvements to ZAMG (Austrian) weather data.  Replaces pull #5526.

**Checklist:**

  - [x] Documentation added/updated in home-assistant/home-assistant.github.io#2014
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New files were added to `.coveragerc`.

Zac-HD committed:
Fast & efficient updates for ZAMG weather data

ZAMG updates on the hour, so instead of checking every half-hour we can
check each minute - only after the observations are taken until
receiving them.

Zac-HD committed:
sensor.zamg: test instead of whitelist for station_id

Zac-HD committed:
Autodetect closest ZAMG station if not given

Zac-HD committed:
ZAMG weather component, based on the sensor

fabaff commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -62,19 +60,31 @@
         vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
-    vol.Required(CONF_STATION_ID):
-        vol.All(cv.string, vol.In(VALID_STATION_IDS)),
+    vol.Optional(CONF_STATION_ID): cv.string,
```
Any reason why we don't use `int`? The station ID is a number.

Zac-HD replied:
I can't speak authoritatively on the Austrian system, but in Australia station IDs may have leading zeros.  Without knowing much more about the schema, I left it as a string.

fabaff replied:
According to the `Stationliste` the IDs have no leading zeros. Most IDs are 11xxx and one exception (8989042) as I'm just looking at the public available information.

fabaff commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -62,19 +60,31 @@
-    station_id = config.get(CONF_STATION_ID)
-    name = config.get(CONF_NAME)
-
+    name = config.get(CONF_NAME, DEFAULT_NAME)
```
The fallback is already handled by `voluptuous`.

fabaff commented on homeassistant/components/weather/zamg.py:
```diff
@@ -0,0 +1,106 @@
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return 'ZAMG {}'.format(self.stationname or '(unknown station)')
```
We should allow the user to set the name with the station name as default.

Zac-HD replied:
This already happens, unless I misunderstand the schema validation.

`ZamgWeather.__init__` sets `self.stationname` to the `stationname` argument (ie `config.get(CONF_NAME)`), and falls back to the station name in the data if this is falsey.  `"(unknown station)"` would only be displayed if this is also missing - ie the data format has changed or something was corrupted.

fabaff replied:
I was comparing the configuration part with the sensor where we use `DEFAULT_NAME`. Looks like we are doing it in different ways for the `owm` and the `bom` weather component.

fabaff replied:
Sorry, I guess that my thoughts in the first comments were more going in the direction not to hard-code the prefix.

fabaff commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -140,40 +145,90 @@ def __init__(self, logger, station_id):
+           'zamg-observation-points')
+    return {e['SYNNR']: tuple(float(e[k].replace(',', '.'))
+                              for k in ['BREITE_DEZI', 'LÄNGR_DEI'])
+            for e in csv.DictReader(requests.get(url).text.splitlines(),
```
Can you please add a timeout to the request?

fabaff commented on homeassistant/components/sensor/zamg.py:
```diff
@@ -140,40 +145,90 @@ def __init__(self, logger, station_id):
+        date, time = self.data.get('update_date'), self.data.get('update_time')
+        if date is not None and time is not None:
+            return datetime.strptime(date + time, '%d-%m-%Y%H:%M').replace(
+                tzinfo=pytz.timezone('Europe/Vienna'))
```
Returning an ISO 8601 timestamp is fine.

Zac-HD replied:
Using a `datetime` object makes it much easier to calculate whether new data is available in the `update` method, so I'd rather leave it.

Zac-HD committed:
Review improvements

fabaff commented:
I got this traceback while running.

```bash  
File "/home/dev/home-assistant/homeassistant/components/sensor/zamg.py", line 76, in setup_platform
    if station_id not in zamg_stations(hass.config.config_dir):
  File "/home/dev/home-assistant/homeassistant/components/sensor/zamg.py", line 212, in zamg_stations
    stations = _get_zamg_stations()
  File "/home/dev/home-assistant/homeassistant/components/sensor/zamg.py", line 200, in _get_zamg_stations
    delimiter=';', quotechar='"')
  File "/home/dev/home-assistant/homeassistant/components/sensor/zamg.py", line 201, in <dictcomp>
    if e['SYNNR'] in capital_stations}
KeyError: 'SYNNR'
```

Zac-HD committed:
Update to new ZAMG schema, add logging

Turns out it wasn't a typo, but rather an upstream schema change.  Added
better error handling to ease diagnosis in case it happens again.

Zac-HD commented:
@fabaff - turns out that ZAMG changed their schema between my commit and your run - I guess we're lucky it happened before merge!  I've therefore updated the function, *and* added some logging to make diagnosis easier if they do it again.

fabaff commented:
If a user chooses to name the sensor 'test123' then the entity_id will currently be `weather.zamg_test123` and the  `friendly_name` "ZAMG test123". `name:` should overwrite the default name completely.

Zac-HD committed:
No hardcoded name
